In [1]:
import numpy as np

In [2]:
%load_ext Cython

## Prime Numbers

In [3]:
def isPrime(N):
    for i in range(2, N):
        if N % i == 0:
            return False
    return True

In [4]:
def isPrime_np(N):
    rem = np.remainder(N, np.arange(2, N))
    return not np.any(rem[:] == 0)

In [5]:
%%cython
cpdef isPrimec(int N):
    cdef int i
    for i in range(2, N):
        if N % i == 0:
            return False
    return True

In [6]:
%%timeit -n 200000
isPrime_np(200)

3.17 µs ± 63.6 ns per loop (mean ± std. dev. of 7 runs, 200000 loops each)


In [7]:
%%timeit -n 200000
isPrime(200)

172 ns ± 62.1 ns per loop (mean ± std. dev. of 7 runs, 200000 loops each)


In [9]:
%%timeit -n 200000
isPrimec(200)

50.1 ns ± 16.8 ns per loop (mean ± std. dev. of 7 runs, 200000 loops each)


## Sorting

In [10]:
import random

In [11]:
def simple_sort(array_):
    array = array_.copy()
    return_ = []
    for i in range(len(array)):
        smallest = min(array)
        index = array.index(smallest)
        return_.append(array[index])
        array.pop(index)
    return return_

In [12]:
def simple_sort_np(array_):
    array = array_.copy()
    return_ = np.zeros((len(array)))
    mx = array.max()
    for i in range(len(array)):
        index = array.argmin()
        return_[i] = array[index]
        array[i] = mx
    return return_

In [13]:
%%cython
cpdef list simple_sortc(list array_):
    cdef list array = array_.copy()
    cdef list return_ = []
    cdef int i, smallest, index
    for i in range(len(array)):
        smallest = min(array)
        index = array.index(smallest)
        return_.append(array[index])
        array.pop(index)
    return return_

In [14]:
arr_np = np.random.randint(1, 100, 100)
arr = list(arr_np)
arr2 = list(arr_np)
print(arr)

[3, 20, 6, 52, 44, 54, 50, 68, 81, 50, 14, 12, 72, 26, 72, 89, 68, 71, 42, 93, 13, 66, 28, 20, 36, 34, 68, 23, 10, 15, 61, 33, 23, 44, 46, 31, 90, 97, 70, 96, 82, 63, 22, 2, 91, 83, 66, 13, 73, 88, 54, 36, 57, 18, 25, 16, 36, 15, 48, 30, 87, 57, 75, 89, 16, 73, 84, 22, 69, 26, 69, 73, 74, 24, 71, 65, 90, 39, 14, 12, 85, 22, 31, 57, 95, 1, 57, 41, 45, 19, 2, 84, 1, 28, 18, 41, 77, 35, 25, 89]


In [15]:
%%timeit -n 10000
simple_sort(arr)

115 µs ± 1.36 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [16]:
%%timeit -n 10000
simple_sort_np(arr_np)

38.7 µs ± 1.68 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [17]:
%%timeit -n 10000
simple_sortc(arr2)

294 µs ± 1.54 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


### 2015 MBP
- `279 µs ± 20.1 µs per loop`
- `103 µs ± 1.1 µs per loop`
- `779 µs ± 37.5 µs per loop`

### 2020 MBA
- `115 µs ± 1.36 µs per loop`
- `38.7 µs ± 1.68 µs per loop`
- `294 µs ± 1.54 µs per loop`